mkPythonDerivation: validate propagatedBuildInputs and buildInputs use a matching python

grahamc opened the pull request:
It isn't so difficult to accidentally mix and match Python packages from different pythons. This can cause fairly mysterious build time and run time errors. This PR adds a trace-only warning:

```
trace: warning: Warning: Python version mismatch in 'hi-bogus':

The Python derivation 'hi-bogus' depends on a Python derivation
named 'foooo', but the two derivations use different versions
of python:

    hi-bogus uses /nix/store/d44wd6n98f93hjr6q1d1phhh1hw7a17d-python3-3.8.8
       foooo uses /nix/store/pp0fp5zds7wqhyx6zqnnxl2aqzf4a51r-python-2.7.18

Suggestion: change hi-bogus's propagatedBuildInputs to use a 'foooo'
built from the same version of Python in /home/grahamc/projects/github.com/NixOS/nixpkgs/test.nix:5:3.
```

which will prevent these mistakes from entering Nixpkgs but only present as a nuisance for users outside of nixpkgs. 

Open questions for me:

- Should this refuse to build?
- Should there be a way to turn it off for specific inputs?
- Should this check apply to other inputs?

_This has been made on behalf of Flox._

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cole-h commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -104,6 +104,45 @@ else
+
+      The Python derivation '${name}' depends on a Python derivation
+      named '${drvName drv}', but the two derivations use different versions
+      of python:
```
Nit:

```suggestion
      of Python:
```

grahamc replied:
:+1:

grahamc commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -104,6 +104,45 @@ else
+          ${leftPadName name (drvName drv)} uses ${python}
+          ${leftPadName (drvName drv) name} uses ${toString drv.pythonModule}
+
+      Suggestion: change ${name}'s ${attrName} to use a '${drvName drv}'
+      built from the same version of Python${optionalLocation}.
+    ''
```
Another good suggestion would be:

```
move the package from ${attrName} to nativeBuildInputs
```

in the case of, example, using python3's black on an python2 project.

cole-h replied:
Maybe have a list of possible solutions?

```
Possible solutions:
  * change ${name}'s ${attrName} to use a '${drvName drv}'
    built from the same version of Python${optionalLocation}.

  * move ${drvName drv} from ${attrName} to nativeBuildInputs.
```

cole-h replied:
Unless you have a way to determine exactly which suggestion should be shown :P

FRidh commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -104,6 +104,49 @@ else
+      # all pythonModules have the pythonModule attribute
+      (drv ? "pythonModule")
+      # Some pythonModules are turned in to a pythonApplication by setting the field to false
+      && (!builtins.isBool drv.pythonModule);
```
in the future we can consider a second passthru attribute so you can do this test always.

grahamc replied:
I was just going to call out this condition and check to see if you thought it was the right condition!

FRidh reviewed (commented):
Good idea!

FRidh commented:
> Should this refuse to build?

In principle I should say yes, because this will lead to mistakes. A potential, but bad case, is where one wants to rely on `wrapPythonPrograms` to add items in `propagatedBuildInputs` to `$PATH`. Note this wouldn't detect it because `pythonModule = false;`.

> Should there be a way to turn it off for specific inputs?

Until someone reports a valid case where it is needed, I don't think we should.

> Should this check apply to other inputs?

This is sufficient I think. `nativeBuildInputs` can indeed contain 2 and 3 mixed. Other cross inputs I think we can ignore.

FRidh commented:
> Should this refuse to build?

Note if it would block to build, it would need an entry in the changelog.

lukegb commented:
nit: propogatedBuildInputs --> propagatedBuildInputs in the PR subject

grahamc commented:
> nit: propogatedBuildInputs --> propagatedBuildInputs in the PR subject

https://logs.nix.samueldr.com/nixos-dev/2021-04-20#4878138;

cole-h commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -104,6 +104,49 @@ else
+          ${leftPadName name (drvName drv)} uses ${python}
+          ${leftPadName (drvName drv) name} uses ${toString drv.pythonModule}
+
+      Possible solutions:
```
Maybe there should be additional explanation that would make it easier to determine which route should be attempted first.

e.g. "If ${drvName drv} provides an executable used during the build, move ${drvName drv}...",
"If ${drvName drv} provides a library used during build, change ${name}'s..."

I'd also make all drv names (including `${name}`) use single quotes; would require some slight rewording to remove the possessive `${name}'s`.

grahamc replied:
What do you think about this?

```
trace: warning: Python version mismatch in 'hi-bogus':

The Python derivation 'hi-bogus' depends on a Python derivation
named 'flake8', but the two derivations use different versions
of Python:

    'hi-bogus' uses /nix/store/d44wd6n98f93hjr6q1d1phhh1hw7a17d-python3-3.8.8
      'flake8' uses /nix/store/pp0fp5zds7wqhyx6zqnnxl2aqzf4a51r-python-2.7.18

Possible solutions for /home/grahamc/projects/github.com/NixOS/nixpkgs/test.nix:5:3:

  * If this is a Python library you depend on, change the reference to 'flake8'
    in the propagatedBuildInputs of 'hi-bogus' to use a 'flake8' built from the same
    version of Python.

  * If 'flake8' provides executables, move the reference to 'flake8' in
    'hi-bogus' from propagatedBuildInputs to nativeBuildInputs.
```

cole-h replied:
Slight wording change in the suggestions:

```
trace: warning: Python version mismatch in 'hi-bogus':

The Python derivation 'hi-bogus' depends on a Python derivation
named 'flake8', but the two derivations use different versions
of Python:

    'hi-bogus' uses /nix/store/d44wd6n98f93hjr6q1d1phhh1hw7a17d-python3-3.8.8
      'flake8' uses /nix/store/pp0fp5zds7wqhyx6zqnnxl2aqzf4a51r-python-2.7.18

Possible solutions:

  * If 'flake8' is a Python library, change its reference in the propagatedBuildInputs
    of 'hi-bogus' to use a 'flake8' built from the same version of Python.

  * If 'flake8' provides executables used while building the package, move the reference
    to 'flake8' in 'hi-bogus' from propagatedBuildInputs to nativeBuildInputs.

at /home/grahamc/projects/github.com/NixOS/nixpkgs/test.nix:5:3
```

cole-h commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -104,6 +104,49 @@ else
+
+    drvName = drv: drv.pname or drv.name;
+    warnMismatch = drv: lib.warn ''
+      Warning: Python version mismatch in '${name}':
```
`Warning:` seems unnecessary; `lib.warn` automatically prepends `warning:` (and `trace:`, but...).

```suggestion
      Python version mismatch in '${name}' detected:
```

grahamc commented:
> Note if it would block to build, it would need an entry in the changelog.

It now blocks the build, @FRidh. Which changelog do you mean? NixOS? Nixpkgs hasn't really had a changelog in years.

FRidh commented:
> It now blocks the build, @FRidh. Which changelog do you mean? NixOS? Nixpkgs hasn't really had a changelog in years.

Yes, NixOS, so it is add least findable without having to resort to `git log` and wade through all the merge commits.

>   * If 'flake8' provides executables, move the reference to 'flake8' in
    'hi-bogus' from propagatedBuildInputs to nativeBuildInputs

- If 'flake8' is used as a tool during the build, move the reference to 'flake8' in
    'hi-bogus' from propagatedBuildInputs to nativeBuildInputs
- If 'flake8' provides executables that need to be available during run-time, pass `makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ flake8 ]}" ];`

grahamc commented:
> If 'flake8' provides executables that need to be available during run-time, pass makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ flake8 ]}" ];

This one is tricky, we don't actually know what the attribute name of the derivation named 'flake8', and I'm not sure how to word this otherwise.

FRidh commented:
I noticed the same when working on error messages on `mk-python-derivation.nix` in https://github.com/NixOS/nixpkgs/pull/102613 and chose to just use the `pname`. To me that seems "good enough". Yes, it will be confusing at times, but in my opinion better than nothing. Otherwise replace the "flake8" there with "<your package>".

grahamc commented:
How about an ofborg check that really wants pname == attrname

grahamc commented:
Current error looks ilke:

```
error: Python version mismatch in 'hi-bogus':

The Python derivation 'hi-bogus' depends on a Python derivation
named 'flake8', but the two derivations use different versions
of Python:

    'hi-bogus' uses /nix/store/d44wd6n98f93hjr6q1d1phhh1hw7a17d-python3-3.8.8
      'flake8' uses /nix/store/pp0fp5zds7wqhyx6zqnnxl2aqzf4a51r-python-2.7.18

Possible solutions:

  * If 'flake8' is a Python library, change the reference to 'flake8'
    in the propagatedBuildInputs of 'hi-bogus' to use a 'flake8' built from the same
    version of Python

  * If 'flake8' is used as a tool during the build, move the reference to
    'flake8' in 'hi-bogus' from propagatedBuildInputs to nativeBuildInputs

  * If 'flake8' provides executables that are called at run time, pass its
    bin path to makeWrapperArgs:

        makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ flake8 ] }" ];

 at /home/grahamc/projects/github.com/NixOS/nixpkgs/test.nix:6:3
```

FRidh commented:
> Current error looks ilke:

Looks great to me!

> How about an ofborg check that really wants pname == attrname

That would also need to consider variants, which would then be forced to fix their `pname`, for example like
```
hdf5-mpi = appendToName "mpi" (hdf5.override { ...
```
That seems like a good idea.

A problematic case would be Python packages. Their `pname` is actually different; the function you're editing now prefixes `pythonX.Y` to its name. An option there is to make that `pname` for example `pythonXYPackages.flake`. But then how do we deal with a dot in a `pname`? This clearly could get complex.

grahamc commented:
Agreed, I don't think it would be reasonable to ensure pname == attrname. It might be a useful lint as part of a "packaging health report", though... Not sure. Something like this would be advisory, not errory.

veprbl commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -104,6 +104,61 @@ else
+
+    drvName = drv: drv.pname or drv.name;
+
+    warnMismatch = drv: let
```
```suggestion
    throwMismatch = drv: let
```

SuperSandro2000 commented:
> How about an ofborg check that really wants pname == attrname

Yes please. or nixpkgs-hammering.

lukegb commented:
Any chance we can land this? The error looks good - I think we might want to file a followup issue against OfBorg for the pname==attrname check.

RaitoBezarius commented:
Can I do something to help get this merged?

Artturin committed:
mkPythonDerivation: validate propagatedBuildInputs use a matching python

mkPythonDerivation: apply checks to buildInputs as well

fixup: add a suggestion about nativeBuildInputs

Clean up the wording, add reasons for the suggestions

switch to throw

Adjust error to clarify build-time vs. run-time

Artturin commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -104,6 +104,61 @@ else
+        len = lib.max (lib.stringLength name) (lib.stringLength against);
+      in lib.strings.fixedWidthString len " " name;
+
+    drvName = drv: drv.pname or drv.name;
```
Perhaps this should just be .name
```

       error: Python version mismatch in 'xpybutil-0.0.6':

       The Python derivation 'xpybutil-0.0.6' depends on a Python derivation
       named 'xpybutil', but the two derivations use different versions
       of Python:

           'xpybutil-0.0.6' uses /nix/store/fdqpyj613dr0v1l1lrzqhzay7sk4xg87-python3-3.10.10
                 'xpybutil' uses /nix/store/v7h4276ij0sj48qwrf4g4cm3pfc86mmw-python3-3.9.16

       Possible solutions:

         * If 'xpybutil' is a Python library, change the reference to 'xpybutil'
           in the propagatedBuildInputs of 'xpybutil-0.0.6' to use a 'xpybutil' built from the same
           version of Python

         * If 'xpybutil' is used as a tool during the build, move the reference to
           'xpybutil' in 'xpybutil-0.0.6' from propagatedBuildInputs to nativeBuildInputs

         * If 'xpybutil' provides executables that are called at run time, pass its
           bin path to makeWrapperArgs:

               makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ xpybutil ] }" ];
```

```diff
diff --git a/pkgs/development/python-modules/xpybutil/default.nix b/pkgs/development/python-modules/xpybutil/default.nix
index cc574c58241..9236df77d84 100644
--- a/pkgs/development/python-modules/xpybutil/default.nix
+++ b/pkgs/development/python-modules/xpybutil/default.nix
@@ -1,4 +1,4 @@
-{ lib, buildPythonPackage, fetchFromGitHub, xcffib, pillow }:
+{ lib, buildPythonPackage, fetchFromGitHub, xcffib, pillow, python39 }:
 
 buildPythonPackage rec {
   pname = "xpybutil";
@@ -13,7 +13,7 @@ buildPythonPackage rec {
   };
 
   # pillow is a dependency in image.py which is not listed in setup.py
-  propagatedBuildInputs = [ pillow xcffib ];
+  propagatedBuildInputs = [ pillow xcffib python39.pkgs.xpybutil ];
 
   propagatedNativeBuildInputs = [ xcffib ];

```

Artturin replied:
changed to .name and the suggestion below to lib.getName

FRidh reviewed (approved):
I have not tested this but the change looks good.

Artturin commented:
```
error: Python version mismatch in 'python3.10-xpybutil-0.0.6':

The Python derivation 'python3.10-xpybutil-0.0.6' depends on a Python derivation
named 'python3.9-i3ipc-2.2.1', but the two derivations use different versions
of Python:

    'python3.10-xpybutil-0.0.6' uses /nix/store/fdqpyj613dr0v1l1lrzqhzay7sk4xg87-python3-3.10.10
        'python3.9-i3ipc-2.2.1' uses /nix/store/v7h4276ij0sj48qwrf4g4cm3pfc86mmw-python3-3.9.16

Possible solutions:

  * If 'python3.9-i3ipc-2.2.1' is a Python library, change the reference to i3ipc
    in the propagatedBuildInputs of 'python3.10-xpybutil-0.0.6' to use a i3ipc built from the same
    version of Python

  * If 'python3.9-i3ipc-2.2.1' is used as a tool during the build, move the reference to
    i3ipc in 'python3.10-xpybutil-0.0.6' from propagatedBuildInputs to nativeBuildInputs

  * If 'python3.9-i3ipc-2.2.1' provides executables that are called at run time, pass its
    bin path to makeWrapperArgs:

        makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ i3ipc ] }" ];
```

grahamc commented:
neat!
